README.md

```diff
@@ -25,27 +31,49 @@ $ jupyter nbextension     list
 
 ```
 $ python -m jupyter_ascending.scripts.make_pair --base examples/test
 ```
 
-Which will create a pair of files: `examples/test.sync.py` and `examples/test.sync.ipynb`. You can read the help for the command to find more information.
+This will create a pair of files: `examples/test.sync.py` and `examples/test.sync.ipynb`.
 
 
 To manually test the ability to sync between a paired python file and a notebook, open up the notebook, and run this python command. You should see the contents of the python file appear in the notebook.
 
 `python -m jupyter_ascending.requests.sync --filename /full/path/to/file.sync.py`
 
+Note that currently Jupyter Ascending expects the jupyter server to be running at `localhost:8888`.
+
+## Security Warning!!
+
+The jupyter-ascending client-server connection is currently completely unauthenticated, even if you have auth enabled on the Jupyter server. This means that, if your jupyter server port is open to the internet, someone could detect that you have jupyter-ascending running, then sync and run arbitrary code on your machine. That's bad!!
+
+For the moment, we recommend only running jupyter-ascending when you're using jupyter locally, or when your jupyter server isn't open to the public internet. For example, we run Jupyter on remote servers, but keep Jupyter accessible only to localhost. Then we use a secure SSH tunnel to do port-forwarding.
+
+Hopefully we can add authentication in the future - it's just rather tricky because there are various forms of auth accepted by Jupyter and not all of them are convenient to integrate into our clients (eg IDE integrations). We welcome contributions here!
+
+## How it works
+
+- your editor calls the jupyter ascending client library with one of a few commands:
+    - sync the code to the notebook (typically on save)
+    - run a cell / run all cells / other commands that should be mapped to a keyboard shortcut
+- the client library assembles a HTTP POST request and sends it to the jupyter server
+- there is a jupyter server extension which accepts HTTP POST requests at `http://[notebook_url]:[notebook_port]/jupyter_ascending`.
+    - Currently `notebook_url:notebook_port` is hardcoded to `localhost:8888`. This should be improved in the future.
+- the server extension matches the request filename to the proper running notebooks and forwards the command along to the notebook plugin
+- a notebook plugin receives the command, and updates the contents of the notebook or executes the requested command.
+
+## Working on a remote server
+
+Because of the client-server architecture, Jupyter Ascending doesn't know or care if the editor and the jupyter server are on the same machine. The client is just sending requests to `http://[notebook_url]:[notebook_port]/jupyter_ascending`, with (hardcoded) default being `http://localhost:8888/jupyter_ascending`. We typically use SSH to forward the remote jupyter port into `localhost:8888`, but you can set up the networking however you like.
```

Comment:
Let's make sure there's at least an option to change the port via an env var?

Reply:
And host obviously

Reply:
I can't think of an easy, clean way to do this. The problem is that if you allow the server addr/port to change, then the kernel extension needs to know what it is (in order to register with the server). And as far as I can tell there's no clean way for the python process in the kernel to get this info. Probably how this should work in an ideal world is that we tunnel the serverextension<->kernelextension comms thru the ZMQSocket that's normally used for these comms, instead of creating our own webserver in the kernel and communicating with that. But I looked into that and it doesn't seem easy.

So we could do a half-measure, like requiring you start the notebook with `JUPYTER_PORT=x JUPYTER_PORT_RETRIES=0 jupyter notebook` (which are already env variables for setting the server port and not allowing it to auto-find a free port), and then accessing the JUPYTER_PORT env variable from the kernel. And also requiring that the kernel and host run on the same machine (so localhost is the right addr). But idk, that's kind of unintuitive to explain to a user?

Ofc it's easy to have env variables on the client side to set the host and port to wherever your server ended up being.

Reply:
I think i just wanted to ensure that, if the server is running on randomhost.com:1234 there was actually a way to configure the client to go to randomhost.com:1234, which I think you're saying is covered by client variables? Then I'm not sure what you meant in the docs by:
```
with (hardcoded) default being `http://localhost:8888/jupyter_ascending
```
I guess I'm still pretty confused?

I thought we moved to a place where we're no longer starting a second web server, so jupyter_ascending client just needs to connect to whatever host and port the jupyter server is running on?

Reply:
Yeah, I can make the client be able to connect to wherever easily. But say the server starts up on :8889 instead of :8888. Then, notebooks won't be able to register themselves, since they're hardcoded to connect to localhost:8888 - and there's no easy way for them to figure out that the server is actually at localhost:8889.

But I guess there's value in adding the client options - connecting to a host other than localhost is fine, and maybe they remapped port 8888 elsewhere eg in a SSH tunnel.

Reply:
Ok, I think you're right that I was making this harder than necessary. There were already env variables set up for this, so I updated the docs to explain how to use them. See the commit called "improve docs around using on a different host:port" for the changes.

Reply:
thanks!